[MRG+1] StratifiedKFold docstring fix

wdevazelhes opened the pull request:
#### Reference Issues/PRs
Fixes #10154.

Commit:
Fix documentation of StratifiedKFold

Commit:
add the same modif to model_selection

agramfort commented on sklearn/cross_validation.py:
```diff
@@ -524,8 +524,10 @@ class StratifiedKFold(_BaseKFold):
-    All the folds have size trunc(n_samples / n_folds), the last one has the
-    complementary.
+    For each class `c`, the first `n_samples_c % n_splits` folds have
+    `n_samples_c // n_splits + 1` samples of class `c`, and other folds have
```
why the +1?

amueller replied:
that doesn't seem to reflect the output that you get in your issue, right?

amueller commented:
Thanks, the docs are pretty obviously wrong, because the number of samples depend on the class distribution. I have to read the code to double check what we actually do (and if it's sensible :-/)

wdevazelhes commented:
My results lengths were in fact the combination of two lenghts (because they were the size of the training set), but I updated the example in the issue so that it is clearer now with the validation set.
Indeed fold `i` is in fact validation set `i`, according to the code.

In addition, below is a more detailed example looking at the number of samples of each class in each fold: 
```python
import numpy as np 
from sklearn.model_selection import StratifiedKFold     
d = np.zeros((15, 2))                           
t = np.hstack([np.zeros(7), np.ones(8)])        
s = StratifiedKFold(n_splits=3)
gen = s.split(d, t)
for i, e in enumerate(gen):                          
    selected = t[e[1]]    
    for c in [0, 1]:         
        print('Fold {}: elems of class {}: {}'.format(i, c, len(selected[selected==c])))
```

Prints:
```
Fold 0 elems of class 0: 3
Fold 0 elems of class 1: 3
Fold 1 elems of class 0: 2
Fold 1 elems of class 1: 3
Fold 2 elems of class 0: 2
Fold 2 elems of class 1: 2
```

So here the change of this PR reflects the output.

As far as I understand this is because StratifiedKFold uses *"individual KFold splitting strategies for each class so as to respect the balance of classes"* (from `StratifiedKFold._make_test_folds`), and according to `KFold` docstring: *"The first `n_samples % n_splits` folds have size `n_samples // n_splits + 1`, other folds have size `n_samples // n_splits`, where `n_samples` is the number of samples."*

wdevazelhes commented:
I think it may be sensible when there are a lot of classes which cardinality is not divisible by `n_splits`, like for instance in Olivetti Faces dataset (40 classes, of 10 elements each): there is a significant difference in validation set sizes:
```python
import numpy as np
from sklearn.model_selection import StratifiedKFold
from sklearn.datasets import fetch_olivetti_faces
dataset = fetch_olivetti_faces()
d, t = dataset.data, dataset.target
s = StratifiedKFold(n_splits=3)  
gen = s.split(d, t)
[e[1].shape[0] for e in gen]  # prints: [160, 120, 120]
 ```

qinhanmin2014 reviewed (approved):
LGTM. I think this reflects the source code. @wdevazelhes Please use double backticks instead of single backtick.

Commit:
double backquotes instead of single

wdevazelhes commented:
@qinhanmin2014 Thanks for the review. I have just put double backticks.

lesteve commented on sklearn/cross_validation.py:
```diff
@@ -524,8 +524,10 @@ class StratifiedKFold(_BaseKFold):
-    complementary.
+    For each class `c`, the first `n_samples_c % n_splits` folds have
+    `n_samples_c // n_splits + 1` samples of class `c`, and other folds have
+    `n_samples_c // n_splits` of them, where `n_samples_c` is the number of
```
You still have single instead of double-backticks in this docstring

wdevazelhes replied:
Ah yes sorry :p will change it

Commit:
add backticks also in model.cross-validation

lesteve commented:
> individual KFold splitting strategies for each class so as to respect the balance of classes" (from StratifiedKFold._make_test_folds), and according to KFold docstring: "The first n_samples % n_splits folds have size n_samples // n_splits + 1, other folds have size n_samples // n_splits, where n_samples is the number of samples."

Is there a way to stay vague and still be informative, i.e. mention that training test sizes may vary across folds or something with a quick explanation as to why ?

My opinion on such implementation details in docstrings is that they tend to rot very quickly. It is so easy to change the code and to forget to update the docstring ...

qinhanmin2014 commented:
> My opinion on such implementation details in docstrings is that they tend to rot very quickly. It is so easy to change the code and to forget to update the docstring ...

Personally, I might not consider it a reason to prevent us from providing details in the doc.

How about mentioning we are doing KFold for each class at the beginning (to serve as a summary of our method before the detail) ? Though precisely, it's a bit different from KFold because we support situations when number of samples in specific class < n_splits.

Commit:
correct flake8 error

lesteve commented:
For me the right level of details is something like: train and test sizes may be different in each fold, with a difference of at most n_classes (not 100 % sure about that this needs to be double-checked). Possibly a little explanation along the lines of you can not at the same time balance class distribution per fold and number of samples in train/test folds.

qinhanmin2014 commented:
> with a difference of at most n_classes (not 100 % sure about that this needs to be double-checked)

I think it's right

> For me the right level of details is something like: train and test sizes may be different in each fold, with a difference of at most n_classes (not 100 % sure about that this needs to be double-checked).

I might write something like:
We use KFold to split each class, so train and test sizes may be different in each fold, with a difference of at most n_classes.
because I still think we need to mention the basic idea of our implementation (KFold for each class). Either solution LGTM from my side.

qinhanmin2014 commented:
ping @lesteve Could you please take some time to figure out an acceptable version to correct the wrong document? From my perspective, both the PR version and your version is acceptable.

qinhanmin2014 commented:
a slight ping @lesteve @jnothman @amueller since the PR is about wrong doc (especially considering we have an email about StratifiedKFold). I think the following versions all LGTM from my side.
(1) current PR (seems more consistent with other functions, e.g. KFold)
(2) Train and test sizes may be different in each fold, with a difference of at most n_classes. (from lesteve)
(3) We use KFold to split each class, so train and test sizes may be different in each fold, with a difference of at most n_classes. (my version from lesteve's comment)

jnothman commented:
I agree with @lesteve that the present PR is too much detail; the reader will not bother to decode what that means for them.

qinhanmin2014 commented:
@wdevazelhes Could you please figure out a version based on the comment from jnothman and lesteve? Thanks.

Commit:
add changes according to github discussion

Commit:
add double backticks

wdevazelhes commented:
Since the rule is not strictly KFold for each class as @qinhanmin2014 highlighted (some particular treatment is applied to non stratifiable classes), I just pushed proposition (2), but tell me if I should go for the third one or an alternative. I agree that the original PR is indeed too much detail.

qinhanmin2014 reviewed (approved):
LGTM. Merging since I believe there's consensus. Thanks @wdevazelhes
